### ISSUE TITLE
custom key [re]mappings ?

### ISSUE THREAD

**Issue by mckellygit** [tomszilagyi/zutty#29]
Hi,
Thank you for this project!  I am excited to explore it.
I have looked but so far cannot find out of there is any way to re-map some key sequences ?
Either through X resources or a custom configuration ?
For example, with urxvt:
keysym.C-BackSpace:  string:\0 ....
with alacritty I can do similar with its yml config.
thx again,
-m

**Comment by tomszilagyi** [comment]
Hi,
There is (curently) no way to alter the keybindings at runtime.
You can of course alter the tables in the source code (look at `src/vterm.cc`).
I know many other terminals provide this feature, but I don't think it is necessary by default.
So I am curious about your usecase, please tell me more about what and why you need to remap.

**Comment by tomszilagyi** [comment]
For the record, current sequences generated by key bindings are documented at https://tomscii.sig7.se/zutty/doc/KEYS.html

**Comment by mckellygit** [comment]
Hi,
Thanks.
Keep in mind I may be a vim/tmux/terminal zealot :-)
But for example I can 'adjust' a small handful of keys with modifiers to get mappings I've enjoyed for years in vim/tmux.
Such as differentiating between <BS>, <S-BS>, <A-BS>, <C-BS>, <A-S-BS>, <C-S-BS>, <C-A-BS> and similarly for <Space> and couple of others.  Without a few binding additions to create pseudo mappings that I handle in tmux/vim these don't often produce different outputs.
Perhaps its been also because I've tried many terminals and wanted the same behavior across all of them.
Which sort of brings up another question, is there a way to tell I'm in zutty - I mean I want TERM to be xterm-256color, but if I knew I was in zutty I could maybe handle things like this in tmux without needing bindings ?
I sometimes see \Eq 2 errors from cursor shape/visibility sequences sent sometimes from the shell (I think) and if I knew I was in zutty I could disable that foo.
thx and really appreciate your work,
-m

**Comment by mckellygit** [comment]
Perhaps checking for ZUTTY_VERSION would do ?
Looking for how to set modifyOtherKeys=2

**Comment by tomszilagyi** [comment]
Yes, I think the best way to determine that you are running within Zutty is to check the existence of the `ZUTTY_VERSION` environment variable. You can also send the ENQ control (`printf "\005"`) to which Zutty will send a unique reply, but this might change in the unlikely case that I find it to cause compatibility issues.

So with modifyOtherKeys set to 2, the keys you listed above all generate different sequences, according to xterm's de-facto standard (as documented). If you are a vim person: newer versions of vim (>= 8.1.2194) set modifyOtherKeys to 2 on startup by default. Here are the sequences that Zutty will generate in this mode (again, according to the documentation I linked above):

- backspace: `DEL`
- Shift-backspace: `CSI 27 ; 2 ; 127 ~`
- Alt-backspace: `CSI 27 ; 3 ; 127 ~`
- Alt-Shift-backspace: `CSI 27 ; 4 ; 127 ~`
- Ctrl-backspace: `CSI 27 ; 5 ; 127 ~`
- Ctrl-Shift-backspace: `CSI 27 ; 6 ; 127 ~`
- Ctrl-Alt-backspace: `CSI 27 ; 7 ; 127 ~`
- Ctrl-Alt-Shift-backspace: `CSI 27 ; 8 ; 127 ~`

You can easily test this yourself by typing e.g., `printf "\e[>4;2m" && cat -vT` followed by the above keypresses.

In summary, since the above is a de-facto standard encoding (introduced by xterm), enabling the terminal application to distinguish between all sorts of funky keybindings, I don't think any customizations are needed.

Let me know if this answers your question / solves your issue.

**Comment by mckellygit** [comment]
Thanks.  I will work on an updated vim/tmux config of my mappings for zutty.
I think we can close this issue as resolved.
take care,
-m

**Comment by mckellygit** [comment]
@tomszilagyi do you support M-Enter ?  I am trying to detect that.

**Comment by tomszilagyi** [comment]
Good point about Alt-Return, this seems to have slipped through...
Fix in latest master.

**Comment by mckellygit** [comment]
Awesome, thanks so much!
I'm getting closer, still tracking down a few of my keys, but perhaps its with tmux.
I'll keep digging.
-m

**Comment by mckellygit** [comment]
Hi,
Do you support S-PageUp/Down ?  I don't seem to see that.  But perhaps X hijacks it for scroll-back scroll ?
thx,
-m

**Comment by tomszilagyi** [comment]
As documented here: https://tomscii.sig7.se/zutty/doc/USAGE.html#User%20interface%20actions
Shift+PageUp/PageDown is captured and triggers the scrollback buffer pager.
It is not captured by X, it is captured by Zutty.

**Comment by mckellygit** [comment]
Thanks.
I was wondering if I had saveLines set to zero and either altScroll on/off, would there be a way for S-PageUp get through ?
Similarly for the scrollWheel up/down ?
And for specifying a custom copy/cut/paste program or way to disable C-S-v / C-S-c and have it get through ?
I've been working with tmux some, trying to help with a modifyOtherKeys=2 mode.
take care,
-m

**Comment by tomszilagyi** [comment]
Sure, if you look at https://github.com/tomszilagyi/zutty/blob/master/src/main.cc#L422
Just remove lines 423 to 442. Then you should receive these keys in your program with the same encoding (if modifyOtherKeys is set to 2).
Personally, I have Ctrl+PageUp/PageDown bound in tmux, so I can choose between "native" scrollback and tmux scrollback.
